<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Configuration &mdash; TRAL 0.3.5 documentation</title>
    
    <link rel="stylesheet" href="_static/tral.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.3.5',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="TRAL 0.3.5 documentation" href="index.html" />
    <link rel="up" title="Installation" href="install.html" />
    <link rel="next" title="Using TRAL in a container or VM" href="install_containerized.html" />
    <link rel="prev" title="p-Value distribution files" href="pvaluefiles.html" />
    <link href='http://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,700|Source+Code+Pro|Armata|IM+Fell+English'
          rel='stylesheet' type='text/css' />
 
    <style type="text/css">
      table.right { float: right; margin-left: 20px; }
      table.right td { border: 1px solid #ccc; }
      
    </style>
    <script type="text/javascript">
      // intelligent scrolling of the sidebar content
      $(window).scroll(function() {
        var sb = $('.sphinxsidebarwrapper');
        var win = $(window);
        var sbh = sb.height();
        var offset = $('.sphinxsidebar').position()['top'];
        var wintop = win.scrollTop();
        var winbot = wintop + win.innerHeight();
        var curtop = sb.position()['top'];
        var curbot = curtop + sbh;
        // does sidebar fit in window?
        if (sbh < win.innerHeight()) {
          // yes: easy case -- always keep at the top
          sb.css('top', $u.min([$u.max([0, wintop - offset - 10]),
                                $(document).height() - sbh - 200]));
        } else {
          // no: only scroll if top/bottom edge of sidebar is at
          // top/bottom edge of window
          if (curtop > wintop && curbot > winbot) {
            sb.css('top', $u.max([wintop - offset - 10, 0]));
          } else if (curtop < wintop && curbot < winbot) {
            sb.css('top', $u.min([winbot - sbh - offset - 20,
                                  $(document).height() - sbh - 200]));
          }
        }
      });
    </script>

  </head>
  <body role="document">
<div class="pageheader">
    <a href="http://www.isb-sib.ch/" class="sib_logo" title="SIB Swiss Institute of Bioinformatics"><img src="_static/sib.png" alt="SIB" class="sib_logo" /></a>
    <a href="index.html">
      <img class="headlogo" src="_static/trallogo.png" alt="TRAL" />
      <span class="headlogo">Tandem Repeat Annotation Library</span>
    </a>
    <div class="headlinks">
        <a href="index.html">Home</a>
        <a href="install.html">Install</a>
        <a href="code_docs.html">Code docs</a>
    </div>
</div>


      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Configuration</a><ul>
<li><a class="reference internal" href="#configs-ini-required-supply-paths-to-external-software-and-data">configs.ini - REQUIRED: Supply paths to external software and data.</a><ul>
<li><a class="reference internal" href="#de-novo-repeat-detection"><em>De novo</em> repeat detection</a></li>
<li><a class="reference internal" href="#build-hmmer-models">Build Hmmer models</a></li>
<li><a class="reference internal" href="#realign-tandem-repeat-units">Realign tandem repeat units</a></li>
<li><a class="reference internal" href="#simulation-of-evolution-in-tandem-repeats">Simulation of evolution in tandem repeats</a></li>
</ul>
</li>
<li><a class="reference internal" href="#configs-ini-optional-change-tral-s-default-behaviour">configs.ini - OPTIONAL: Change TRAL&#8217;s default behaviour</a><ul>
<li><a class="reference internal" href="#define-the-default-filter-behaviour">Define the default filter behaviour.</a></li>
<li><a class="reference internal" href="#creation-of-new-repeat-instances">Creation of new Repeat instances</a></li>
<li><a class="reference internal" href="#statistical-significance-calculation-and-models-of-repeat-evolution">Statistical significance calculation and models of repeat evolution</a></li>
<li><a class="reference internal" href="#creation-of-new-repeat-list-instances">Creation of new Repeat list instances</a></li>
<li><a class="reference internal" href="#restrict-hmmer-model-size">Restrict Hmmer model size.</a></li>
</ul>
</li>
<li><a class="reference internal" href="#logging-ini-optional">logging.ini - OPTIONAL</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="pvaluefiles.html"
                        title="previous chapter">p-Value distribution files</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="install_containerized.html"
                        title="next chapter">Using TRAL in a container or VM</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/configure.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="configuration">
<span id="configure"></span><h1>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h1>
<p>Here you see how to configure TRAL after installation. All configuration files are located
in your home directory:</p>
<div class="highlight-python"><div class="highlight"><pre>~/.tral/
</pre></div>
</div>
<div class="section" id="configs-ini-required-supply-paths-to-external-software-and-data">
<h2>configs.ini - REQUIRED: Supply paths to external software and data.<a class="headerlink" href="#configs-ini-required-supply-paths-to-external-software-and-data" title="Permalink to this headline">¶</a></h2>
<p>In configs.ini you can set the paths to all external software and data used in TRAL, as well
as change default thresholds and behaviours.</p>
<div class="section" id="de-novo-repeat-detection">
<h3><em>De novo</em> repeat detection<a class="headerlink" href="#de-novo-repeat-detection" title="Permalink to this headline">¶</a></h3>
<p>Enlist all <em>de novo</em> detection algorithms you wish to use. All other algorithms are ignored
by default.</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detection]]
        AA = HHrepID, T-REKS, TRUST, XSTREAM  # List of default de novo detectors on amino acid data that you wish to use.
        DNA = PHOBOS, XSTREAM  #List of default de novo detectors on nucleic data that you wish to use.
</pre></div>
</div>
<p>If the binaries or executable scripts of any <em>de novo</em> detection algorithm are not in the
system path, set the absolute paths in config.ini. Details for each detector (e.g. to create
executable scripts) are detailed in <a class="reference internal" href="install_external.html#install-external"><span>Installation of external software</span></a>.</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        PHOBOS = phobos
        HHrepID = hhrepid_64
        HHrepID_dummyhmm = /path/to/home/.tral/data/hhrepid/dummyHMM.hmm
        TRUST = TRUST
        TRUST_substitutionmatrix = /path/to/TRUST/Align/BLOSUM50  #Recommended: Use matrices packages shipped with TRUST
        XSTREAM = XSTREAM
        [...]
</pre></div>
</div>
</div>
<div class="section" id="build-hmmer-models">
<h3>Build Hmmer models<a class="headerlink" href="#build-hmmer-models" title="Permalink to this headline">¶</a></h3>
<p>If <em>hmmbuild</em> is not in your system path, define the absolute path:</p>
<div class="highlight-python"><div class="highlight"><pre>[hmm]
    hmmbuild = path/to/hmmbuild
</pre></div>
</div>
</div>
<div class="section" id="realign-tandem-repeat-units">
<h3>Realign tandem repeat units<a class="headerlink" href="#realign-tandem-repeat-units" title="Permalink to this headline">¶</a></h3>
<p>At current, tandem repeat units are realigned with Mafft&#8217;s ginsi global aligner. This is
required for example if you annotate tandem repeats from profile models. If <em>ginsi</em> is
not in your system path, define the absolute path:</p>
<div class="highlight-python"><div class="highlight"><pre>[repeat]
    ginsi = /path/to/ginsi
</pre></div>
</div>
<p>Additionally, all detected tandem repeats (either from profile models or <em>denovo</em> can be realigned with the indel-aware proPIP alignment algorithm.
To use proPIP you need to have Castor (incl. aligner) which ships with proPIP. If <em>Castor</em> is not in your system path, define the absolute path:</p>
<div class="highlight-python"><div class="highlight"><pre>[repeat]
    Castor = /path/to/Castor
    [[castor_parameter]]
        rate_distribution = constant # either constant or gamma
</pre></div>
</div>
</div>
<div class="section" id="simulation-of-evolution-in-tandem-repeats">
<h3>Simulation of evolution in tandem repeats<a class="headerlink" href="#simulation-of-evolution-in-tandem-repeats" title="Permalink to this headline">¶</a></h3>
<p>If <em>alfsim</em> is not in your system path, set the absolute path:</p>
<div class="highlight-python"><div class="highlight"><pre>[repeat]
        alfsim = path/to/alfsim
</pre></div>
</div>
</div>
</div>
<div class="section" id="configs-ini-optional-change-tral-s-default-behaviour">
<h2>configs.ini - OPTIONAL: Change TRAL&#8217;s default behaviour<a class="headerlink" href="#configs-ini-optional-change-tral-s-default-behaviour" title="Permalink to this headline">¶</a></h2>
<div class="section" id="define-the-default-filter-behaviour">
<h3>Define the default filter behaviour.<a class="headerlink" href="#define-the-default-filter-behaviour" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>[filter]
    [[basic]]
        tag = basic_filter
        [[[dict]]]
            [[[[pvalue]]]]
                func_name = pvalue
                score = phylo_gap01  # score
                threshold = 0.1  # p-Value cut-off
            [[[[n_effective]]]]
                func_name = attribute
                attribute = n_effective  # attribute of tandem repeat, e.g. the repeat unit length l_effective
                type = min
                threshold = 1.9  # p-Value cut-off
</pre></div>
</div>
</div>
<div class="section" id="creation-of-new-repeat-instances">
<h3>Creation of new Repeat instances<a class="headerlink" href="#creation-of-new-repeat-instances" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>[repeat]
    scoreslist = phylo_gap01,  # score
    calc_score = False  # Is the score calculated?
    calc_pvalue = False # Is the pvalue calculated?
    precision = 10
    ginsi = /path/to/ginsi  # Path to the mafft global aligner ginsi.
    Castor = Castor
    [[castor_parameter]]
        rate_distribution = constant # either constant or gamma
    alfsim = alfsim
</pre></div>
</div>
</div>
<div class="section" id="statistical-significance-calculation-and-models-of-repeat-evolution">
<h3>Statistical significance calculation and models of repeat evolution<a class="headerlink" href="#statistical-significance-calculation-and-models-of-repeat-evolution" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>[repeat_score]
    evolutionary_model = lg  # All models need to be available in /path/to/TandemRepeats/tandemrepeats/data/paml/
    [[indel]]
        indelRatePerSite = 0.01  # What magnitude is the indel rate compared to the substitution rate?
        ignore_gaps = True  # Shall trailing gaps be ignored / not penalised?
        gaps = row_wise
        zipf = 1.821
    [[optimisation]]
        start_min = 0.5
        start_max = 1.5
        nIteration = 14
    [[K80]]
        kappa = 2.59
    [[TN93]]
        alpha_1 = 0.3
        alpha_2 = 0.4
        beta = 0.7
[[score_calibration]]
    scoreslist=phylo_gap01,
    save_calibration = False
    precision = 10
</pre></div>
</div>
</div>
<div class="section" id="creation-of-new-repeat-list-instances">
<h3>Creation of new Repeat list instances<a class="headerlink" href="#creation-of-new-repeat-list-instances" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>[repeat_list]
    output_characteristics = begin, msa_original, l_effective, n_effective, repeat_region_length, divergence, pvalue
    model = phylo_gap01
</pre></div>
</div>
</div>
<div class="section" id="restrict-hmmer-model-size">
<h3>Restrict Hmmer model size.<a class="headerlink" href="#restrict-hmmer-model-size" title="Permalink to this headline">¶</a></h3>
<p>Set the maximum size of HMM for which the Viterbi algorithm is performed <em>l_effective_max</em> e.g.
to ensure viable run-times on your system:</p>
<div class="highlight-python"><div class="highlight"><pre>[hmm]
    l_effective_max = 50
</pre></div>
</div>
</div>
</div>
<div class="section" id="logging-ini-optional">
<h2>logging.ini - OPTIONAL<a class="headerlink" href="#logging-ini-optional" title="Permalink to this headline">¶</a></h2>
<p>In this file, you can define the level of debugging per module (DEBUG, INFO, WARNING), and
the format of the debugging message. Defaults to WARNING. The path to the file needs to be
defined as</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">logging.config</span>
<span class="n">logging</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">fileConfig</span><span class="p">(</span><span class="s2">&quot;path/to/your/home/.tral/logging.ini&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>


  </body>
</html>