<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Configuration &mdash; TRAL 0.3.0 documentation</title>
    
    <link rel="stylesheet" href="_static/tral.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.3.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="TRAL 0.3.0 documentation" href="index.html" />
    <link href='http://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,700|Source+Code+Pro|Armata|IM+Fell+English'
          rel='stylesheet' type='text/css' />
 
    <style type="text/css">
      table.right { float: right; margin-left: 20px; }
      table.right td { border: 1px solid #ccc; }
      
    </style>
    <script type="text/javascript">
      // intelligent scrolling of the sidebar content
      $(window).scroll(function() {
        var sb = $('.sphinxsidebarwrapper');
        var win = $(window);
        var sbh = sb.height();
        var offset = $('.sphinxsidebar').position()['top'];
        var wintop = win.scrollTop();
        var winbot = wintop + win.innerHeight();
        var curtop = sb.position()['top'];
        var curbot = curtop + sbh;
        // does sidebar fit in window?
        if (sbh < win.innerHeight()) {
          // yes: easy case -- always keep at the top
          sb.css('top', $u.min([$u.max([0, wintop - offset - 10]),
                                $(document).height() - sbh - 200]));
        } else {
          // no: only scroll if top/bottom edge of sidebar is at
          // top/bottom edge of window
          if (curtop > wintop && curbot > winbot) {
            sb.css('top', $u.max([wintop - offset - 10, 0]));
          } else if (curtop < wintop && curbot < winbot) {
            sb.css('top', $u.min([winbot - sbh - offset - 20,
                                  $(document).height() - sbh - 200]));
          }
        }
      });
    </script>

  </head>
  <body>
<div class="pageheader">
    <a href="http://www.isb-sib.ch/" class="sib_logo" title="SIB Swiss Institute of Bioinformatics"><img src="_static/sib.png" alt="SIB" class="sib_logo" /></a>
    <a href="index.html">
      <img class="headlogo" src="_static/trallogo.png" alt="TRAL" />
      <span class="headlogo">Tandem Repeat Annotation Library</span>
    </a>
    <div class="headlinks">
        <a href="index.html">Home</a>
        <a href="installation.html">Install</a>
        <a href="code_docs.html">Code docs</a>
    </div>
</div>


      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Configuration</a><ul>
<li><a class="reference internal" href="#configs-ini-required-changes">configs.ini: Required changes</a><ul>
<li><a class="reference internal" href="#de-novo-repeat-detection"><em>De novo</em> repeat detection</a></li>
<li><a class="reference internal" href="#build-hmmer-models">Build Hmmer models</a></li>
<li><a class="reference internal" href="#realign-tandem-repeat-units">Realign tandem repeat units</a></li>
</ul>
</li>
<li><a class="reference internal" href="#configs-ini-defaults">configs.ini: Defaults</a><ul>
<li><a class="reference internal" href="#parameters-of-the-default-filter">Parameters of the default filter</a></li>
<li><a class="reference internal" href="#default-behaviour-when-new-repeat-instances-are-created">Default behaviour when new Repeat instances are created</a></li>
<li><a class="reference internal" href="#default-parameters-for-the-model-of-repeat-evolution-used-for-statistical-significance-calculation">Default parameters for the model of repeat evolution used for statistical significance calculation</a></li>
<li><a class="reference internal" href="#default-behaviour-when-new-repeat-list-instances-are-created">Default behaviour when new Repeat list instances are created</a></li>
</ul>
</li>
<li><a class="reference internal" href="#logging-ini">logging.ini</a></li>
<li><a class="reference internal" href="#p-value-distribution-files">p-Value distribution files</a></li>
</ul>
</li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/configure.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="configuration">
<span id="configure"></span><h1>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h1>
<p>Here you see how to configure TRAL after installation. All configuration files are located
in your home directory:</p>
<div class="highlight-python"><div class="highlight"><pre>~/.tral/
</pre></div>
</div>
<div class="section" id="configs-ini-required-changes">
<h2>configs.ini: Required changes<a class="headerlink" href="#configs-ini-required-changes" title="Permalink to this headline">¶</a></h2>
<p>In configs.ini you need to set the paths to all external software used in TRAL.</p>
<div class="section" id="de-novo-repeat-detection">
<h3><em>De novo</em> repeat detection<a class="headerlink" href="#de-novo-repeat-detection" title="Permalink to this headline">¶</a></h3>
<p>Here, define the paths to all de novo detection algorithms you wish to use. The makeup of the
scripts or binaries needed is detailed in <a class="reference internal" href="install_external.html#install-external"><em>Installation of external software</em></a>.
Only paths for algorithms defined under [[repeat_detection]] are used, all other paths are
ignored.</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detection]]
        AA = *List of default de novo detectors on amino acid data that you wish to use*
        DNA = *List of default de novo detectors on nucleic data that you wish to you*
    [[repeat_detector_path]]
        PHOBOS = path/to/PHOBOS
        HHrepID = path/to/HHREPID
        HHrepID_dummyhmm = path/to/HHREPID_HMM_DB
        TRUST = path/to/TRUST
        TRUST_substitutionmatrix = path/to/substitution_matrix/e.g.BLOSUM50 #Recommended: Use matrices packages with TRUST
        XSTREAM = path/to/XSTREAM
        [...]
</pre></div>
</div>
</div>
<div class="section" id="build-hmmer-models">
<h3>Build Hmmer models<a class="headerlink" href="#build-hmmer-models" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>[hmm]
    hmmbuild = path/to/hmmbuild
    lDMax = 50  # maximum size of HMM for which the Viterbi algorithm is performed
</pre></div>
</div>
</div>
<div class="section" id="realign-tandem-repeat-units">
<h3>Realign tandem repeat units<a class="headerlink" href="#realign-tandem-repeat-units" title="Permalink to this headline">¶</a></h3>
<p>At current, tandem repeat units are realigned with Mafft&#8217;s ginsi global aligner. If tandem
repeat unit realignment is required, for example, if you annotate tandem repeats from profile models,
set the path to ginsi:</p>
<div class="highlight-python"><div class="highlight"><pre>[repeat]
    ginsi = /path/to/ginsi  # Path to the mafft global aligner ginsi.
</pre></div>
</div>
</div>
</div>
<div class="section" id="configs-ini-defaults">
<h2>configs.ini: Defaults<a class="headerlink" href="#configs-ini-defaults" title="Permalink to this headline">¶</a></h2>
<p>In configs.ini you can additionally change TRAL&#8217;s default behaviour.</p>
<div class="section" id="parameters-of-the-default-filter">
<h3>Parameters of the default filter<a class="headerlink" href="#parameters-of-the-default-filter" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>[filter]
    [[basic]]
        tag = basic_filter
        [[[dict]]]
            [[[[pValue]]]]
                func_name = pValue
                score = phylo_gap01  # score
                threshold = 0.1  # p-Value cut-off
            [[[[nD]]]]
                func_name = attribute
                attribute = nD  # attribute of tandem repeat, e.g. the repeat unit length lD
                type = min
                threshold = 1.9  # p-Value cut-off
</pre></div>
</div>
</div>
<div class="section" id="default-behaviour-when-new-repeat-instances-are-created">
<h3>Default behaviour when new Repeat instances are created<a class="headerlink" href="#default-behaviour-when-new-repeat-instances-are-created" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>[repeat]
    scoreslist = phylo_gap01,  # score
    calc_score = False  # Is the score calculated?
    calc_pValue = False # Is the pValue calculated?
    precision = 10
    GINSI = /path/to/ginsi  # Path to the mafft global aligner ginsi.
</pre></div>
</div>
</div>
<div class="section" id="default-parameters-for-the-model-of-repeat-evolution-used-for-statistical-significance-calculation">
<h3>Default parameters for the model of repeat evolution used for statistical significance calculation<a class="headerlink" href="#default-parameters-for-the-model-of-repeat-evolution-used-for-statistical-significance-calculation" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>[repeat_score]
    evolutionary_model = lg  # All models need to be available in /path/to/TandemRepeats/tandemrepeats/data/paml/
    [[indel]]
        indelRatePerSite = 0.01  # What magnitude is the indel rate compared to the substitution rate?
        ignore_gaps = True  # Shall trailing gaps be ignored / not penalised?
        gaps = row_wise
        zipf = 1.821
    [[optimisation]]
        start_min = 0.5
        start_max = 1.5
        nIteration = 14
    [[K80]]
        kappa = 2.59
    [[TN93]]
        alpha_1 = 0.3
        alpha_2 = 0.4
        beta = 0.7
[[score_calibration]]
    scoreslist=phylo_gap01,
    save_calibration = False
    precision = 10
</pre></div>
</div>
</div>
<div class="section" id="default-behaviour-when-new-repeat-list-instances-are-created">
<h3>Default behaviour when new Repeat list instances are created<a class="headerlink" href="#default-behaviour-when-new-repeat-list-instances-are-created" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>[repeat_list]
    msa_original = True  # Is msa_original calculated?
    lD = True  # Is lD calculated?
    nD = True  # Is nD calculated?
    sequence_length = True  # Is the sequence_length calculated?
    pValue = phylo_gap01
    begin = True  # Is the position in sequence calculated?
</pre></div>
</div>
</div>
</div>
<div class="section" id="logging-ini">
<h2>logging.ini<a class="headerlink" href="#logging-ini" title="Permalink to this headline">¶</a></h2>
<p>In this file, you can define the level of debugging per module (DEBUG, INFO, WARNING), and
the format of the debugging message. Defaults to WARNING. The path to the file needs to be
defined as</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">logging.config</span>
<span class="n">logging</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">fileConfig</span><span class="p">(</span><span class="s">&quot;path/to/your/home/.tral/logging.ini&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="p-value-distribution-files">
<h2>p-Value distribution files<a class="headerlink" href="#p-value-distribution-files" title="Permalink to this headline">¶</a></h2>
<p>In order to calculate the p-Value of tandem repeat scores, available p-Value distributions
need to be downloaded and placed in <em>./tral/data/pValue</em>:</p>
<div class="highlight-python"><div class="highlight"><pre>cd ~/.tral/data/pValue
svn checkout https://github.com/elkeschaper/tral/trunk/tral/data/pValue .
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>


  </body>
</html>