<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Installation of external software &mdash; TRAL 0.3.5 documentation</title>
    
    <link rel="stylesheet" href="_static/tral.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.3.5',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="TRAL 0.3.5 documentation" href="index.html" />
    <link href='http://fonts.googleapis.com/css?family=Source+Sans+Pro:300,400,700|Source+Code+Pro|Armata|IM+Fell+English'
          rel='stylesheet' type='text/css' />
 
    <style type="text/css">
      table.right { float: right; margin-left: 20px; }
      table.right td { border: 1px solid #ccc; }
      
    </style>
    <script type="text/javascript">
      // intelligent scrolling of the sidebar content
      $(window).scroll(function() {
        var sb = $('.sphinxsidebarwrapper');
        var win = $(window);
        var sbh = sb.height();
        var offset = $('.sphinxsidebar').position()['top'];
        var wintop = win.scrollTop();
        var winbot = wintop + win.innerHeight();
        var curtop = sb.position()['top'];
        var curbot = curtop + sbh;
        // does sidebar fit in window?
        if (sbh < win.innerHeight()) {
          // yes: easy case -- always keep at the top
          sb.css('top', $u.min([$u.max([0, wintop - offset - 10]),
                                $(document).height() - sbh - 200]));
        } else {
          // no: only scroll if top/bottom edge of sidebar is at
          // top/bottom edge of window
          if (curtop > wintop && curbot > winbot) {
            sb.css('top', $u.max([wintop - offset - 10, 0]));
          } else if (curtop < wintop && curbot < winbot) {
            sb.css('top', $u.min([winbot - sbh - offset - 20,
                                  $(document).height() - sbh - 200]));
          }
        }
      });
    </script>

  </head>
  <body role="document">
<div class="pageheader">
    <a href="http://www.isb-sib.ch/" class="sib_logo" title="SIB Swiss Institute of Bioinformatics"><img src="_static/sib.png" alt="SIB" class="sib_logo" /></a>
    <a href="index.html">
      <img class="headlogo" src="_static/trallogo.png" alt="TRAL" />
      <span class="headlogo">Tandem Repeat Annotation Library</span>
    </a>
    <div class="headlinks">
        <a href="index.html">Home</a>
        <a href="install.html">Install</a>
        <a href="code_docs.html">Code docs</a>
    </div>
</div>


      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Installation of external software</a><ul>
<li><a class="reference internal" href="#sequence-profile-model-generation">Sequence profile model generation</a></li>
<li><a class="reference internal" href="#alignment-of-tandem-repeat-units">Alignment of tandem repeat units</a></li>
<li><a class="reference internal" href="#currently-integrated-detectors">Currently integrated detectors</a><ul>
<li><a class="reference internal" href="#hhrepid">HHrepID</a></li>
<li><a class="reference internal" href="#phobos">PHOBOS</a></li>
<li><a class="reference internal" href="#tred">TRED</a></li>
<li><a class="reference internal" href="#t-reks">T-REKS</a></li>
<li><a class="reference internal" href="#trf">TRF</a></li>
<li><a class="reference internal" href="#trust">TRUST</a></li>
<li><a class="reference internal" href="#xstream">XSTREAM</a></li>
</ul>
</li>
<li><a class="reference internal" href="#not-yet-integrated-software">Not yet integrated software</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/install_external.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="installation-of-external-software">
<span id="install-external"></span><h1>Installation of external software<a class="headerlink" href="#installation-of-external-software" title="Permalink to this headline">¶</a></h1>
<p>Here, we explain how to install external software packages, such as <em>de novo</em> tandem repeat
detectors. Important: Once a software is installed, the path to either the binary or the
executable shell script needs to be indicated in the TRAL configuration file <a class="reference internal" href="configure.html#configure"><span>config.ini</span></a>.</p>
<div class="section" id="sequence-profile-model-generation">
<h2>Sequence profile model generation<a class="headerlink" href="#sequence-profile-model-generation" title="Permalink to this headline">¶</a></h2>
<p>At current, only support for HMMER is integrated <a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/24/6/807.long">published in Bioinformatics (2008)</a> (<a class="reference external" href="http://hmmer.janelia.org/">Installation instructions</a>).
TRAL searches for HMMER&#8217;s <em>hmmbuild</em> in the system path by default (modify in <a class="reference internal" href="configure.html#configure"><span>config.ini</span></a>):</p>
<div class="highlight-python"><div class="highlight"><pre>[hmm]
    hmmbuild = hmmbuild
</pre></div>
</div>
<p>If <em>hmmbuild</em> is not in your system path, set the absolute path:</p>
<div class="highlight-python"><div class="highlight"><pre>[hmm]
    hmmbuild = path/to/hmmbuild
</pre></div>
</div>
</div>
<div class="section" id="alignment-of-tandem-repeat-units">
<span id="mafft"></span><h2>Alignment of tandem repeat units<a class="headerlink" href="#alignment-of-tandem-repeat-units" title="Permalink to this headline">¶</a></h2>
<p>Currently, MAFFT is the advised tool for (re-)alignment of the tandem repeat units to each other (<a class="reference external" href="http://mafft.cbrc.jp/alignment/software/">Installation instructions</a>).
TRAL searches for MAFFT&#8217;s <em>ginsi</em> in the system path by default (modify in <a class="reference internal" href="configure.html#configure"><span>config.ini</span></a>):</p>
<div class="highlight-python"><div class="highlight"><pre>[repeat]
    ginsi = ginsi
</pre></div>
</div>
<p>If <em>ginsi</em> is not in your system path, set the absolute path:</p>
<div class="highlight-python"><div class="highlight"><pre>[repeat]
    ginsi = path/to/ginsi
</pre></div>
</div>
</div>
<div class="section" id="currently-integrated-detectors">
<span id="install-denovo"></span><h2>Currently integrated detectors<a class="headerlink" href="#currently-integrated-detectors" title="Permalink to this headline">¶</a></h2>
<table border="1" class="docutils">
<colgroup>
<col width="35%" />
<col width="65%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Software</th>
<th class="head">License</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>HHrepID</td>
<td>CC-BY-NC-2.0</td>
</tr>
<tr class="row-odd"><td>PHOBOS</td>
<td>Non-commercial</td>
</tr>
<tr class="row-even"><td>TRED</td>
<td>Non-commercial</td>
</tr>
<tr class="row-odd"><td>T-REKS</td>
<td>Non-commercial?</td>
</tr>
<tr class="row-even"><td>TRF</td>
<td>Unlimited use</td>
</tr>
<tr class="row-odd"><td>TRUST</td>
<td>?</td>
</tr>
<tr class="row-even"><td>XSTREAM</td>
<td>Non-commercial</td>
</tr>
</tbody>
</table>
<div class="section" id="hhrepid">
<h3>HHrepID<a class="headerlink" href="#hhrepid" title="Permalink to this headline">¶</a></h3>
<p>HHrepID is a profile self alignment <em>de novo</em> amino acid tandem repeat detection software
<a class="reference external" href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002195">published e.g. in PLOS Computational Biology (2011)</a> (<a class="reference external" href="http://toolkit.tuebingen.mpg.de/hhrepid">Installation instructions</a>).
TRAL searches for the executable binary <em>hhrepid64</em> in the system path by default (modify in <a class="reference internal" href="configure.html#configure"><span>config.ini</span></a>):</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        HHrepID = hhrepid_64
</pre></div>
</div>
<p>If the executable <em>hhrepid</em> is not in your system path, set the absolute path:</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        HHrepID = path/to/hhrepid
</pre></div>
</div>
<p>Also, you need to supply a null hmm file for using HHrepID. A dummy file is located in your home/.tral.
Supply the path to your null hmm file of choice:</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        HHrepID_dummyhmm = /path/to/home/.tral/data/dummyHMM.hmm
</pre></div>
</div>
</div>
<div class="section" id="phobos">
<h3>PHOBOS<a class="headerlink" href="#phobos" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="http://www.ruhr-uni-bochum.de/ecoevo/cm/cm_phobos.htm">PHOBOS</a>  is an unpublished <em>k</em>-mer based <em>de novo</em> DNA tandem repeat detection software.
TRAL searches for the executable <em>phobos</em> in the system path by default (modify in <a class="reference internal" href="configure.html#configure"><span>config.ini</span></a>):</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        PHOBOS = phobos
</pre></div>
</div>
<p>If <em>phobos</em> is not in your system path, set the absolute path:</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        PHOBOS = path/to/phobos
</pre></div>
</div>
</div>
<div class="section" id="tred">
<h3>TRED<a class="headerlink" href="#tred" title="Permalink to this headline">¶</a></h3>
<p>TRED is a sequence self alignment <em>de novo</em> amino acid tandem repeat detection software
<a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/23/2/e30.short">published in Bioinformatics (2007)</a> (The software is available on request).
TRAL searches for the executable <em>tred</em> in the system path by default (modify in <a class="reference internal" href="configure.html#configure"><span>config.ini</span></a>):</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        TRED = tred
</pre></div>
</div>
<p>If <em>phobos</em> is not in your system path, set the absolute path:</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        TRED = path/to/tred
</pre></div>
</div>
</div>
<div class="section" id="t-reks">
<h3>T-REKS<a class="headerlink" href="#t-reks" title="Permalink to this headline">¶</a></h3>
<p>T-REKS is a <em>k</em>-mer based <em>de novo</em> DNA/AA tandem repeat detection software
<a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/25/20/2632.short">published in Bioinformatics (2009)</a> (<a class="reference external" href="http://bioinfo.montp.cnrs.fr/?r=t-reks">Installation instructions</a>).
Create an executable text file T-REKS with the following content:</p>
<div class="highlight-python"><div class="highlight"><pre>#!/bin/sh
java -jar /my/path/to/T-Reks.jar &quot;$@&quot;
</pre></div>
</div>
<p>If you place this text file in your systems path, TRAL finds it by default
(modify in <a class="reference internal" href="configure.html#configure"><span>config.ini</span></a>):</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        T-REKS = T-REKS
</pre></div>
</div>
<p>If you did not place T-REKS in your system path or named it differently, set the absolute
path:</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        T-REKS = path/to/T-REKS
</pre></div>
</div>
</div>
<div class="section" id="trf">
<h3>TRF<a class="headerlink" href="#trf" title="Permalink to this headline">¶</a></h3>
<p>TRF is a <em>k</em>-mer based self alignment <em>de novo</em> DNA tandem repeat detection software
<a class="reference external" href="http://nar.oxfordjournals.org/content/27/2/573.full">published in Nucleic Acids Research (1999)</a> (<a class="reference external" href="http://tandem.bu.edu/trf/trf.html">Installation instructions</a>).
TRAL searches for the executable <em>trf</em> in the system path by default (modify in <a class="reference internal" href="configure.html#configure"><span>config.ini</span></a>):</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        TRF = trf
</pre></div>
</div>
<p>If <em>trf</em> is not in your system path, set the absolute path:</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        TRF = path/to/trf
</pre></div>
</div>
</div>
<div class="section" id="trust">
<h3>TRUST<a class="headerlink" href="#trust" title="Permalink to this headline">¶</a></h3>
<p>TRUST is a sequence self alignment <em>de novo</em> amino acid tandem repeat detection software
<a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/20/suppl_1/i311.short">published in Bioinformatics (2004)</a> (<a class="reference external" href="http://www.ibi.vu.nl/programs/trustwww/">Installation instructions</a>).</p>
<p>Create an executable text file TRUST with the following content (you can amend the java
memory consumption restrictions):</p>
<div class="highlight-python"><div class="highlight"><pre>#!/bin/sh
java -Xmx30G -cp /my/path/to/TRUST/1.0.0/Align nl.vu.cs.align.SelfSimilarity &quot;$@&quot;
</pre></div>
</div>
<p>If you place this text file in your systems path, TRAL finds it by default
(modify in <a class="reference internal" href="configure.html#configure"><span>config.ini</span></a>):</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        TRUST = TRUST
</pre></div>
</div>
<p>If you did not place TRUST in your system path or named it differently, set the absolute
path:</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        TRUST = path/to/TRUST
</pre></div>
</div>
<p>Also, you need to supply a substitution matrix for using TRUST (it ships with several substitution matrices).
Supply the path of your favourite substitution matrix:</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        TRUST_substitutionmatrix = /path/to/TRUST/Align/BLOSUM50
</pre></div>
</div>
</div>
<div class="section" id="xstream">
<span id="id2"></span><h3>XSTREAM<a class="headerlink" href="#xstream" title="Permalink to this headline">¶</a></h3>
<p>XSTREAM is a <em>k</em>-mer based <em>de novo</em> DNA/AA tandem repeat detection software
<a class="reference external" href="http://www.biomedcentral.com/1471-2105/8/382/">published in BMC Bioinformatics (2007)</a> (<a class="reference external" href="http://jimcooperlab.mcdb.ucsb.edu/xstream/download.jsp">Installation instructions</a>).</p>
<p>Create an executable text file XSTREAM with the following content:</p>
<div class="highlight-python"><div class="highlight"><pre>#!/bin/sh
java -jar /my/path/to/xstream.jar &quot;$@&quot;
</pre></div>
</div>
<p>If you place this text file in your systems path, TRAL finds it by default
(modify in <a class="reference internal" href="configure.html#configure"><span>config.ini</span></a>):</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        XSTREAM = XSTREAM
</pre></div>
</div>
<p>If you did not place XSTREAM in your system path or named it differently, set the absolute
path:</p>
<div class="highlight-python"><div class="highlight"><pre>[sequence]
    [[repeat_detector_path]]
        XSTREAM = path/to/XSTREAM
</pre></div>
</div>
</div>
</div>
<div class="section" id="not-yet-integrated-software">
<h2>Not yet integrated software<a class="headerlink" href="#not-yet-integrated-software" title="Permalink to this headline">¶</a></h2>
<p>There is a large number of tandem repeat detection software for which TRAL does not provide
parsers. However, theses parsers are easily manually added to <a class="reference internal" href="code_docs.html#code-docs"><span>sequence.repeat_detection_io</span></a>.
Please file an issue on the <a class="reference external" href="https://github.com/acg-team/tandemrepeats/issues">tracker</a>.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>


  </body>
</html>